bloom-works/coram-uk-child-welfare-playbook
# Implement Cypress with Accessibility testing

## PR Description/Summary

This change is addressing the need for Cypress and accessibility testing of this project. Modeling the [Civic Actions Accessibility playbook guide](https://accessibility.civicactions.com/playbook/automated-testing), we have added automated JavaScript front end testing using [Cypress](https://docs.cypress.io/guides/overview/why-cypress) in combination with [Axe core](https://github.com/dequelabs/axe-core), an accessibility testing engine for websites.

The accessibility tests run in combination with the cypress tests, using the `npx cypress run` or `npx cypress open` commands.

### What was changed

- Updated the Cypress testing framework, along with some specs that already existed
- Installed Axe core for accessibility testing
- README doc updated with developer instructions for running Cypress tests

### How to check change

- Checkout branch and build project
- Make sure the server is running at `http://localhost:8000/`
- Run `npx cypress run` to run and verify accessibility tests in headless mode or Run `npx cypress open` to use Cypress' UI mode
- Creation and configuration of new tests can be implemented in `cypress/e2e` folder
- Cypress e2e configurations can be found in `cypress.config.tsx` and compiler options are located in `cypress/tsconfig.json`
- For testing purposes, the `rules` in `cypress/e2e/accessibility.cy.tsx` can be commented out so the accessibility tests will fail. 

### Additional Notes

** Since we don't have any Coram pages built yet, the pages are showing accessibility errors such as not having an alt-tag or their is a color-contrast issue. Since Adam is still designing the pages and we haven't started building yet, the current tests I have in place in the spec file, `cypress/e2e/accessibility.cy.tsx`, are disabling specific accessibility errors. We will re-enable them once we start building out the Coram playbook. 

** Cypress automatically creates a video when it's run in headless mode, I have disabled this feature in the `cypress.congif.ts` file so it's not created during the GitHub workflow. To re-enable it in your local env, change the value of `video` to `true` or comment it out.
